yamux/src/frame.rs

```diff
@@ -13,89 +13,144 @@ mod io;
 
 use futures::future::Either;
 use header::{Data, GoAway, Header, Ping, StreamId, WindowUpdate};
-use std::{convert::TryInto, num::TryFromIntError};
+use std::{convert::TryInto, fmt::Debug, marker::PhantomData, num::TryFromIntError};
+use zerocopy::{AsBytes, ByteSlice, ByteSliceMut, Ref};
 
 pub use io::FrameDecodeError;
 pub(crate) use io::Io;
 
-/// A Yamux message frame consisting of header and body.
-#[derive(Clone, Debug, PartialEq, Eq)]
+use self::header::HEADER_SIZE;
+
+/// A Yamux message frame consisting of header and body in a single buffer
+#[derive(Clone, Debug, Eq, PartialEq)]
 pub struct Frame<T> {
-    header: Header<T>,
-    body: Vec<u8>,
+    buffer: Vec<u8>,
+    _marker: std::marker::PhantomData<T>,
+}
+
+impl<T> Default for Frame<T> {
+    fn default() -> Self {
+        Self {
+            buffer: Vec::new(),
+            _marker: PhantomData,
+        }
+    }
 }
 
 impl<T> Frame<T> {
-    pub fn new(header: Header<T>) -> Self {
-        Frame {
-            header,
-            body: Vec::new(),
+    pub(crate) fn new(buffer: Vec<u8>) -> Self {
+        Self {
+            buffer,
+            _marker: std::marker::PhantomData,
         }
     }
 
-    pub fn header(&self) -> &Header<T> {
-        &self.header
+    pub(crate) fn from_header(header: Header<T>) -> Self {
+        let mut buffer = vec![0; HEADER_SIZE];
+        header.write_to(&mut buffer).expect("write_to success");
+        Self::new(buffer)
+    }
+
+    fn make_parsed_frame<B: ByteSlice>(
+        header: Ref<B, Header<T>>,
+        body: B,
+    ) -> Result<ParsedFrame<B, T>, io::FrameDecodeError> {
+        let frame = ParsedFrame { header, body };
+        let version = frame.header.version().val();
+        if version != 0 {
+            Err(FrameDecodeError::Header(crate::HeaderDecodeError::Version(
+                version,
+            )))
+        } else {
+            frame.header.tag().map(|_| frame).map_err(|e| e.into())
+        }
     }
 
-    pub fn header_mut(&mut self) -> &mut Header<T> {
-        &mut self.header
+    pub(crate) fn parse(&self) -> Result<ParsedFrame<&[u8], T>, io::FrameDecodeError> {
+        let (header, body) = Ref::new_from_prefix(&self.buffer[..]).expect("construct a valid Ref");
+        Self::make_parsed_frame(header, body)
+    }
+
+    pub(crate) fn parse_mut(&mut self) -> Result<ParsedFrame<&mut [u8], T>, io::FrameDecodeError> {
+        let (header, body) =
+            Ref::new_from_prefix(&mut self.buffer[..]).expect("construct a valid Ref");
+        Self::make_parsed_frame(header, body)
+    }
+
+    pub fn into_buffer(self) -> Vec<u8> {
+        self.buffer
+    }
+
+    pub fn buffer(&self) -> &[u8] {
+        &self.buffer
+    }
```

Comment:
Can we move these further down such that Git shows them as a diff to the previous `body()` and `into_body` functions?

Reply:
ok